### PR TITLE
Task/vs/slcore 495 manage telemetry from core

### PR DESCRIPTION
**Pull request by vojtech-suchy-sonarsource**

### CODE REVIEW

**Review by henryju — COMMENTED** [review]
I made already a few comments last week, but now the PR needs to be rebased, this would make the review easier.

**Review by henryju — CHANGES_REQUESTED** [review]

File: rpc-protocol/src/main/java/org/sonarsource/sonarlint/core/rpc/protocol/client/telemetry/TelemetryLiveAttributesResponse.java [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+    this.usesConnectedMode = usesConnectedMode;
+    this.useSonarCloud = useSonarCloud;
+    this.nodeVersion = nodeVersion;
+    this.devNotificationsDisabled = devNotificationsDisabled;
+    this.nonDefaultEnabledRules = nonDefaultEnabledRules;
+    this.defaultDisabledRules = defaultDisabledRules;
+    this.additionalAttributes = additionalAttributes;
+  }
+
+  public boolean isUsesConnectedMode() {
```
[the remark this answers is not in the document]
Comment (vojtech-suchy-sonarsource) [reply rpc-protocol/src/main/java/org/sonarsource/sonarlint/core/rpc/protocol/client/telemetry/TelemetryLiveAttributesResponse.java:77]: Now reported in https://sonarsource.atlassian.net/browse/SLCORE-628

File: rpc-protocol/src/main/java/org/sonarsource/sonarlint/core/rpc/protocol/client/telemetry/TelemetryLiveAttributesResponse.java [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+    return usesConnectedMode;
+  }
+
+  public boolean isUseSonarCloud() {
```
[the remark this answers is not in the document]
Comment (vojtech-suchy-sonarsource) [reply rpc-protocol/src/main/java/org/sonarsource/sonarlint/core/rpc/protocol/client/telemetry/TelemetryLiveAttributesResponse.java:81]: Now reported in https://sonarsource.atlassian.net/browse/SLCORE-628

File: rpc-protocol/src/main/java/org/sonarsource/sonarlint/core/rpc/protocol/client/telemetry/TelemetryLiveAttributesResponse.java [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+    return useSonarCloud;
+  }
+
+  @Nullable
+  public String getNodeVersion() {
+    return nodeVersion;
+  }
+
+  public boolean isDevNotificationsDisabled() {
```
[the remark this answers is not in the document]
Comment (vojtech-suchy-sonarsource) [reply rpc-protocol/src/main/java/org/sonarsource/sonarlint/core/rpc/protocol/client/telemetry/TelemetryLiveAttributesResponse.java:90]: Now reported in https://sonarsource.atlassian.net/browse/SLCORE-628

File: rpc-protocol/src/main/java/org/sonarsource/sonarlint/core/rpc/protocol/client/telemetry/TelemetryLiveAttributesResponse.java [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+    return devNotificationsDisabled;
+  }
+
+  public List<String> getNonDefaultEnabledRules() {
```
[the remark this answers is not in the document]
Comment (vojtech-suchy-sonarsource) [reply rpc-protocol/src/main/java/org/sonarsource/sonarlint/core/rpc/protocol/client/telemetry/TelemetryLiveAttributesResponse.java:94]: Now reported in https://sonarsource.atlassian.net/browse/SLCORE-628

File: rpc-protocol/src/main/java/org/sonarsource/sonarlint/core/rpc/protocol/client/telemetry/TelemetryLiveAttributesResponse.java [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+    return nonDefaultEnabledRules;
+  }
+
+  public List<String> getDefaultDisabledRules() {
```
[the remark this answers is not in the document]
Comment (vojtech-suchy-sonarsource) [reply rpc-protocol/src/main/java/org/sonarsource/sonarlint/core/rpc/protocol/client/telemetry/TelemetryLiveAttributesResponse.java:98]: Now reported in https://sonarsource.atlassian.net/browse/SLCORE-628

File: rpc-protocol/src/main/java/org/sonarsource/sonarlint/core/rpc/protocol/backend/telemetry/TelemetryRpcService.java [hunk 6]
```diff
@@ -20,12 +20,51 @@
...
+  void addReportedRules(AddReportedRulesParams params);
+
+  @JsonNotification
+  void addQuickFixAppliedForRule(AddQuickFixAppliedForRule params);
+
+  @JsonNotification
+  void helpAndFeedbackLinkClicked(HelpAndFeedbackClickedParams params);
+
+  @JsonNotification
+  void stop();
```
[the remark this answers is not in the document]
Comment (vojtech-suchy-sonarsource) [reply rpc-protocol/src/main/java/org/sonarsource/sonarlint/core/rpc/protocol/backend/telemetry/TelemetryRpcService.java]: Addressed in [a63d13f](https://github.com/SonarSource/sonarlint-core/pull/781/commits/a63d13f315ca41eab6710a91422be3ecc0574b85)

File: rpc-protocol/src/main/java/org/sonarsource/sonarlint/core/rpc/protocol/client/telemetry/AddQuickFixAppliedForRule.java [hunk 7]
```diff
@@ -17,7 +17,16 @@
  * along with this program; if not, write to the Free Software Foundation,
  * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
  */
-@ParametersAreNonnullByDefault
-package org.sonarsource.sonarlint.core.telemetry;
+package org.sonarsource.sonarlint.core.rpc.protocol.client.telemetry;
 
-import javax.annotation.ParametersAreNonnullByDefault;
+public class AddQuickFixAppliedForRule {
```
[the remark this answers is not in the document]
Comment (vojtech-suchy-sonarsource) [reply rpc-protocol/src/main/java/org/sonarsource/sonarlint/core/rpc/protocol/client/telemetry/AddQuickFixAppliedForRule.java]: Addressed in [a63d13f](https://github.com/SonarSource/sonarlint-core/pull/781/commits/a63d13f315ca41eab6710a91422be3ecc0574b85)

File: telemetry/src/main/java/org/sonarsource/sonarlint/core/telemetry/TelemetryService.java [hunk 8]
```diff
@@ -0,0 +1,261 @@
...
+      LOG.info("Telemetry disabled by a system property");
+      return;
+    }
+    this.telemetryLocalStorageManager.tryUpdateAtomically(storage -> storage.setInitialNewCodeFocus(initializeParams.isFocusOnNewCode()));
+    var initialDelay = Integer.parseInt(System.getProperty("sonarlint.internal.telemetry.initialDelay", "1"));
+    scheduledExecutor.scheduleWithFixedDelay(this::upload, initialDelay, TELEMETRY_UPLOAD_DELAY, MINUTES);
+  }
+
+  private void upload() {
+    var clientTelemetryPayload = getTelemetryLiveAttributes();
```
[the remark this answers is not in the document]
Comment (vojtech-suchy-sonarsource) [reply telemetry/src/main/java/org/sonarsource/sonarlint/core/telemetry/TelemetryService.java]: Addressed in [a63d13f](https://github.com/SonarSource/sonarlint-core/pull/781/commits/a63d13f315ca41eab6710a91422be3ecc0574b85)

File: telemetry/src/main/java/org/sonarsource/sonarlint/core/telemetry/TelemetryService.java [hunk 9]
```diff
@@ -0,0 +1,261 @@
...
+      telemetryManager.uploadLazily(clientTelemetryPayload);
+    }
+  }
+
+  public GetStatusResponse getStatus() {
+    return new GetStatusResponse(isEnabled());
+  }
+
+  public void enableTelemetry() {
+    var clientTelemetryPayload = getTelemetryLiveAttributes();
```
[the remark this answers is not in the document]
Comment (vojtech-suchy-sonarsource) [reply telemetry/src/main/java/org/sonarsource/sonarlint/core/telemetry/TelemetryService.java]: Addressed in [a63d13f](https://github.com/SonarSource/sonarlint-core/pull/781/commits/a63d13f315ca41eab6710a91422be3ecc0574b85)

File: telemetry/src/main/java/org/sonarsource/sonarlint/core/telemetry/TelemetryService.java [hunk 10]
```diff
@@ -0,0 +1,261 @@
...
+    if (Objects.nonNull(clientTelemetryPayload)) {
+      telemetryManager.enable(clientTelemetryPayload);
+    }
+  }
+
+  public void disableTelemetry() {
+    var clientTelemetryPayload = getTelemetryLiveAttributes();
```
[the remark this answers is not in the document]
Comment (vojtech-suchy-sonarsource) [reply telemetry/src/main/java/org/sonarsource/sonarlint/core/telemetry/TelemetryService.java]: Addressed in [a63d13f](https://github.com/SonarSource/sonarlint-core/pull/781/commits/a63d13f315ca41eab6710a91422be3ecc0574b85)

File: telemetry/src/test/java/org/sonarsource/sonarlint/core/telemetry/TelemetryHttpClientTests.java [hunk 11]
```diff
@@ -64,7 +62,8 @@ void opt_out() {
     sonarqubeMock.stubFor(delete("/")
       .willReturn(aResponse()));
 
-    underTest.optOut(new TelemetryLocalStorage(), attributes);
+    var telemetryPayload = new TelemetryLiveAttributesResponse(true, false, null, false, emptyList(), emptyList(), emptyMap());
```
[the remark this answers is not in the document]
Comment (vojtech-suchy-sonarsource) [reply telemetry/src/test/java/org/sonarsource/sonarlint/core/telemetry/TelemetryHttpClientTests.java]: Addressed in [a63d13f](https://github.com/SonarSource/sonarlint-core/pull/781/commits/a63d13f315ca41eab6710a91422be3ecc0574b85)

File: telemetry/src/test/java/org/sonarsource/sonarlint/core/telemetry/TelemetryHttpClientTests.java [hunk 12]
```diff
@@ -78,7 +77,8 @@ void upload() {
     sonarqubeMock.stubFor(post("/")
       .willReturn(aResponse()));
 
-    underTest.upload(new TelemetryLocalStorage(), attributes);
+    var telemetryPayload = new TelemetryLiveAttributesResponse(true, false, null, false, emptyList(), emptyList(), emptyMap());
```
[the remark this answers is not in the document]
Comment (vojtech-suchy-sonarsource) [reply telemetry/src/test/java/org/sonarsource/sonarlint/core/telemetry/TelemetryHttpClientTests.java]: Addressed in [a63d13f](https://github.com/SonarSource/sonarlint-core/pull/781/commits/a63d13f315ca41eab6710a91422be3ecc0574b85)

File: telemetry/src/test/java/org/sonarsource/sonarlint/core/telemetry/TelemetryHttpClientTests.java [hunk 13]
```diff
@@ -92,7 +92,8 @@ void should_not_crash_when_cannot_upload() {
     sonarqubeMock.stubFor(post("/")
       .willReturn(aResponse().withStatus(500)));
 
-    underTest.upload(new TelemetryLocalStorage(), attributes);
+    var telemetryPayload = new TelemetryLiveAttributesResponse(true, false, null, false, emptyList(), emptyList(), emptyMap());
```
[the remark this answers is not in the document]
Comment (vojtech-suchy-sonarsource) [reply telemetry/src/test/java/org/sonarsource/sonarlint/core/telemetry/TelemetryHttpClientTests.java]: Addressed in [a63d13f](https://github.com/SonarSource/sonarlint-core/pull/781/commits/a63d13f315ca41eab6710a91422be3ecc0574b85)

**Review by henryju — APPROVED** [review]
Well done!

**Pull request merged**